server/user_db.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes"
+
+	pb "github.com/googleapis/gapic-showcase/server/genproto"
+	"google.golang.org/genproto/protobuf/field_mask"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func NewUserDb() UserDb {
+	return &userDb{
+		uid:   &uniqID{},
+		token: NewTokenGenerator(),
+		mu:    sync.Mutex{},
+		keys:  map[string]dbIndex{},
+		users: []userEntry{},
+	}
+}
+
+type UserDb interface {
+	Create(u *pb.User) (*pb.User, error)
+	Get(name string) (*pb.User, error)
+	Update(u *pb.User, f *field_mask.FieldMask) (*pb.User, error)
+	Delete(name string) error
+	List(page_size int32, token string) (*pb.ListUsersResponse, error)
+}
+
+type userEntry struct {
+	user    *pb.User
+	deleted bool
+}
+
+type dbIndex int
+
+type userDb struct {
+	uid   *uniqID
+	token TokenGenerator
+	mu    sync.Mutex
```

Comment:
`mu` protects `keys` and `users` right? In that case, add new line before `mu`. Idiomatically mutex is the first member in the "group" it's protecting.

Reply:
Thanks! Done!